File: java/coding-interview/src/com/mayfly/interview/dynamicProgramming/leetcode/LongestCommonSubsequence.java
package com.mayfly.interview.dynamicProgramming.leetcode;

import java.util.Arrays;

public class LongestCommonSubsequence {

    public static int solution(String text1, String text2) {
        int[][] table = new int[text2.length()][text1.length()];

        for (int i = 0; i < text2.length(); i++) {
            for (int j = 0; j < text1.length(); j++) {
                if (text2.charAt(i) == text1.charAt(j)) {
                    // carry from diagonal row + 1
                    if (i > 0 && j > 0) table[i][j] = table[i-1][j-1] + 1;
                    // else just = 1
                    else table[i][j] = 1;
                } else {
                    int left = 0, up = 0;
                    if (j > 0) left = table[i][j-1];
                    if (i > 0) up = table[i-1][j];
                    table[i][j] = Math.max(left, up);
                }
            }
        }

        //printTable(table);
        return table[text2.length()-1][text1.length()-1];
    }

    public static void printTable(int[][] table) {
        for (int[] t : table) {
            System.out.println(Arrays.toString(t));
        }
    }

    public static void run() {
        //System.out.println(solution("abcde", "ace")); //3
        //System.out.println(solution("cdfed", "zxdye")); //2
        System.out.println(solution("abcba", "abcbcba")); //5
        System.out.println(solution("pmjghexybyrgzczy", "hafcdqbgncrcbihkd")); //4
    }
}

File: python/coding-interview/leetcode/medium/longest_common_subsequence.py
def solution_recursion(text1, text2):
	def recurse(i, j):
		if i >= len(text1) or j >= len(text2):
			return 0
		elif text1[i] == text2[j]:
			return 1 + recurse(i+1, j+1)
		else:
			return max(recurse(i+1, j), recurse(i, j+1))

	return recurse(0, 0)


def solution(text1, text2):
	table = [[0 for _ in range(len(text2)+1)] for _ in range(len(text1)+1)]

	for i in range(1, len(text1)+1):
		for j in range(1, len(text2)+1):
			a1, a2 = text1[i-1], text2[j-1]
			if a1 == a2:
				table[i][j] = 1 + table[i-1][j-1]
			else:
				table[i][j] = max(table[i-1][j], table[i][j-1])
	
	return table[i][j]


	
assert solution_recursion("abcde", "ace") == 3
assert solution_recursion("abc", "abc") == 3
assert solution_recursion("abc", "def") == 0
assert solution_recursion("efg", "abesfh") == 2

assert solution("abcde", "ace") == 3
assert solution("abc", "abc") == 3
assert solution("abc", "def") == 0
assert solution("efg", "abesfh") == 2
